linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -698,29 +695,68 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
       tcaptureValues: List[PreTransform])(cont: PreTransCont)(
       implicit scope: Scope, pos: Position): TailRec[Tree] = {
 
-    val thisType = if (arrow) NoType else AnyType
+    /* Process captures.
+     *
+     * This is different than normal param replacement for two reasons:
+     *
+     * - We introduce bindings for the capture values. This allows to eliminate
+     *   captures that are not used (but preserve side effects).
+     * - If the capture value is a VarRef, we give the capture param the exact
+     *   same name. This is to help the FunctionEmitter eliminate IIFEs.
+     */
+    val (captureParamLocalDefs, newCaptureParamDefsAndRepls, captureValueBindings) = (for {
+      (paramDef, tcaptureValue) <- captureParams.zip(tcaptureValues)
+    } yield {
+      val ParamDef(ident @ LocalIdent(paramName), originalName, ptpe, mutable) = paramDef
+
+      assert(!mutable)
+
+      val (newName, newOriginalName) = tcaptureValue match {
+        case PreTransLocalDef(LocalDef(_, false, ReplaceWithVarRef(captureName, _, _))) =>
+          (captureName, originalNameForFresh(paramName, originalName, captureName))
+        case _ =>
+          freshLocalName(paramName, originalName, mutable)
+      }
+      val replacement = ReplaceWithVarRef(newName, newSimpleState(Unused), None)
+      val localDef = LocalDef(RefinedType(ptpe), mutable, replacement)
+      val localIdent = LocalIdent(newName)(ident.pos)
+      val newParamDef = ParamDef(localIdent, newOriginalName, ptpe, mutable)(paramDef.pos)
 
-    val (allNewParamsWithUse, newBody) = transformIsolatedBody(None, thisType,
-        captureParams ++ params ++ restParam, AnyType, body)
+      /* Note that the binding will never create a fresh name for a
+       * ReplaceWithVarRef. So this will not put our name alignment at risk.
```

Comment:
Should we add a comment where this decision and taken, to state that this particular property is relied upon over here?

Alternatively, is there a fundamental reason for which it would never do that?

Reply:
I have attempted to add a comment. The main reason is just that its not necessary... :shrug: